### ISSUE TITLE
EISDIR error due to filename with a single number

### ISSUE THREAD

**Issue by kurayama** [yarnpkg/yarn#3966]
**Do you want to request a *feature* or report a *bug*?**

Bug

**What is the current behavior?**
Having a file named 0, 1, 2 or 3 breaks install when using native dependencies.

**If the current behavior is a bug, please provide the steps to reproduce.**

```bash
> touch 1
> yarn add bcrypt
```

Errors with:
```bash
yarn add v0.27.5
info No lockfile found.
[1/4] Resolving packages...
[2/4] Fetching packages...
error An unexpected error occurred: "EISDIR: illegal operation on a directory, open '/Users/foobar/Library/Caches/Yarn/v1/npm-abbrev-0.0.0-c73b17e6-e97a-414a-b3e1-8bb9f2f265a5-1500486460913'".
info If you think this is a bug, please open a bug report with the information provided in "/private/tmp/biz/yarn-error.log".
info Visit https://yarnpkg.com/en/docs/cli/add for documentation about this command.
```

**What is the expected behavior?**
Install dependencies normally

**Please mention your node.js, yarn and operating system version.**
Node v7.10, yarn v0.27.5 and OSX 10.12.5

**Comment by BYK** [comment]
Can't repro on latest master. Can you share your `package.json` file?

**Comment by kurayama** [comment]
@BYK I've updated how to reproduce, you need to have a C dependency like grpc or bcrypt.

**Comment by andrewhao** [comment]
Fails for me:

    $ touch 1
    $ yarn install

Node 8.1.4
Yarn 0.27.5

```
{
  "name": "landing-pages",
  "version": "1.0.0",
  "description": "Stitch Fix Landing Pages",
  "main": "index.js",
  "scripts": {
    "test": "karma start",
    "build": "webpack --config webpack.config.js && jekyll build",
    "build-minified": "webpack -p --config webpack.config.js && jekyll build --config='_config.yml,_config_production.yml,_config_production_assets.yml'",
    "serve": "webpack --config webpack.config.js && jekyll serve",
    "serve-minified": "webpack -p --config webpack.config.js && jekyll serve --config='_config.yml,_config_production.yml,_config_production_assets.yml'",
    "start": "npm run serve"
  },
  "repository": "git+https://github.com/stitchfix/landing-pages.git",
  "keywords": [],
  "author": "Stitch Fix",
  "license": "UNLICENSED",
  "bugs": {
    "url": "https://github.com/stitchfix/landing-pages/issues"
  },
  "homepage": "https://github.com/stitchfix/landing-pages#readme",
  "devDependencies": {
    "babel-core": "^6.7.7",
    "babel-loader": "^6.2.4",
    "babel-preset-es2015": "^6.6.0",
    "babel-preset-react": "^6.5.0",
    "codeclimate-test-reporter": "^0.4.1",
    "karma-coverage": "^1.1.1",
    "react": "^15.0.2",
    "react-addons-update": "^15.0.2",
    "react-dom": "^15.0.2",
    "webpack": "^2.2.1"
  },
  "dependencies": {
    "css-loader": "^0.26.2",
    "extract-text-webpack-plugin": "^2.0.0",
    "install": "^0.8.7",
    "jasmine-core": "^2.5.2",
    "jquery": "^3.1.1",
    "js-cookie": "^2.1.3",
    "karma": "^1.5.0",
    "karma-babel-preprocessor": "^6.0.1",
    "karma-chrome-launcher": "^2.0.0",
    "karma-cli": "^1.0.1",
    "karma-coverage": "^1.1.1",
    "karma-jasmine": "^1.1.0",
    "karma-jasmine-ajax": "^0.1.13",
    "karma-jasmine-jquery": "^0.1.1",
    "karma-phantomjs-launcher": "^1.0.4",
    "karma-webpack": "^2.0.2",
    "kss": "^3.0.0-beta.18",
    "kss-webpack-plugin": "^1.2.0",
    "lodash": "^4.17.4",
    "node-sass": "^4.5.0",
    "postcss-loader": "^1.3.3",
    "sass-loader": "^6.0.2",
    "uglify-js": "^2.8.12",
    "uglifyjs-webpack-plugin": "^0.3.0",
    "webpack-dev-server": "^2.4.1",
    "yaml-loader": "^0.4.0"
  },
  "engines": {
    "node": "7.x"
  }
}
```

**Comment by bytheway875** [comment]
Also, it looks like it specifically fails when looking for a 0.0.0 version of a package in the cache when the dependency version is just "1":

Got the EISDIR error for packages which and abbrev.

```"nopt@2 || 3", nopt@3.x, nopt@~3.0.0, nopt@~3.0.1:
  version "3.0.6"
  resolved "https://registry.yarnpkg.com/nopt/-/nopt-3.0.6.tgz#c6465dbf08abcd4db359317f79ac68a646b28ff9"
  dependencies:
    abbrev "1"

node-gyp@^3.3.1:
  version "3.6.2"
  resolved "https://registry.yarnpkg.com/node-gyp/-/node-gyp-3.6.2.tgz#9bfbe54562286284838e750eac05295853fa1c60"
  dependencies:
    fstream "^1.0.0"
    glob "^7.0.3"
    graceful-fs "^4.1.2"
    minimatch "^3.0.2"
    mkdirp "^0.5.0"
    nopt "2 || 3"
    npmlog "0 || 1 || 2 || 3 || 4"
    osenv "0"
    request "2"
    rimraf "2"
    semver "~5.3.0"
    tar "^2.0.0"
    which "1"
```

**Comment by BYK** [comment]
Duplicate of #4013. 

I know this was reported earlier but #4013 has clear repro steps so gonna use that for tracking. Thanks for the report!

**Comment by mccricardo** [comment]
Souldn't #4013 be closed instead, since this is older?

**Comment by mccricardo** [comment]
Was going to ask just that. Thank you.

**Comment by BYK** [comment]
Fix submitted: #4088
